channeldb/waitingproof.go

```diff
@@ -3,106 +3,261 @@ package channeldb
-		s.cache[proof.Key()] = struct{}{}
-		return nil
-	}); err != nil && err != ErrWaitingProofNotFound {
-		return nil, err
+		db: db,
 	}
 
 	return s, nil
 }
 
-// Add adds new waiting proof in the storage.
-func (s *WaitingProofStore) Add(proof *WaitingProof) error {
-	if _, ok := s.cache[proof.Key()]; ok {
-		return ErrWaitingProofAlreadyExist
-	}
+// AddLocalProof adds a new local channel proof to the database.
+func (s *WaitingProofStore) AddLocalProof(msg *lnwire.AnnounceSignatures) error {
+	shortChanID := msg.ShortChannelID.ToUint64()
+	return s.db.Update(func(tx *bolt.Tx) error {
+		// Get or create the bucket.
+		bucket, err := tx.CreateBucketIfNotExists(waitingProofsBucketKey)
+		if err != nil {
+			return err
+		}
+
+		// Get bucket for this ShortChannelID.
+		var channelBucketKey [8]byte
+		binary.BigEndian.PutUint64(channelBucketKey[:], shortChanID)
+		channelBucket, err := bucket.CreateBucketIfNotExists(channelBucketKey[:])
+		if err != nil {
+			return err
+		}
 
-	return s.db.Batch(func(tx *bolt.Tx) error {
-		var err error
-		var b bytes.Buffer
+		// But the local proof into this channel bucket.
+		return putLocalProof(channelBucket, msg)
+	})
+}
 
+// AddRemoteProof adds a new remote channel proof to the database.
+func (s *WaitingProofStore) AddRemoteProof(msg *lnwire.AnnounceSignatures) error {
+	shortChanID := msg.ShortChannelID.ToUint64()
+	return s.db.Update(func(tx *bolt.Tx) error {
 		// Get or create the bucket.
 		bucket, err := tx.CreateBucketIfNotExists(waitingProofsBucketKey)
 		if err != nil {
 			return err
 		}
 
-		// Encode the objects and place it in the bucket.
-		if err := proof.Encode(&b); err != nil {
+		// Get bucket for this ShortChannelID.
+		var channelBucketKey [8]byte
+		binary.BigEndian.PutUint64(channelBucketKey[:], shortChanID)
+		channelBucket, err := bucket.CreateBucketIfNotExists(channelBucketKey[:])
+		if err != nil {
 			return err
 		}
 
-		key := proof.Key()
-		if err := bucket.Put(key[:], b.Bytes()); err != nil {
+		// Put the remote proof into this channel bucket.
+		return putRemoteProof(channelBucket, msg)
+	})
+}
+
+// AddRemotePeer adds the provided remote peer id to the proof for
+// the channel with the provided ShortChannelID.
+func (s *WaitingProofStore) AddRemotePeer(id lnwire.ShortChannelID,
+	remotePeer *btcec.PublicKey) error {
+	shortChanID := id.ToUint64()
+	return s.db.Update(func(tx *bolt.Tx) error {
+		// Get or create the bucket.
+		bucket, err := tx.CreateBucketIfNotExists(waitingProofsBucketKey)
+		if err != nil {
 			return err
 		}
 
-		s.cache[proof.Key()] = struct{}{}
-		return nil
+		// Get bucket for this ShortChannelID.
+		var channelBucketKey [8]byte
+		binary.BigEndian.PutUint64(channelBucketKey[:], shortChanID)
+		channelBucket, err := bucket.CreateBucketIfNotExists(channelBucketKey[:])
+		if err != nil {
+			return err
+		}
+
+		// Put the peer into this channel bucket.
+		return putPeer(channelBucket, remotePeer)
```

Comment:
yeah, _almost_ enough. by fetching the channel info for that particular channel ID we have the public keys of the two peers of the channel, but I couldn't figure out a clever way of figuring out _which_ of the two is the remote, and which is the local node. 

If there's an elegant way of fetching this missing information, this could be simplified, and we wouldn't need to store the peer id in the database.